File: greencoderefactor-java/src/main/java/uclm/esi/alarcos/greenteam/greencoderefactor_java/transformers/sonar/r1854/UnusedAssignmentTransformer.java
package uclm.esi.alarcos.greenteam.greencoderefactor_java.transformers.sonar.r1854;

import com.github.javaparser.ast.CompilationUnit;
import com.github.javaparser.ast.Node;
import com.github.javaparser.ast.body.MethodDeclaration;
import com.github.javaparser.ast.expr.*;
import com.github.javaparser.ast.stmt.*;
import com.github.javaparser.ast.visitor.ModifierVisitor;
import com.github.javaparser.ast.visitor.Visitable;
import uclm.esi.alarcos.greenteam.greencoderefactor_java.transformers.BaseTransformer;

import java.util.*;

/**
 * Transformer Sonar R1854:
 */
public class UnusedAssignmentTransformer extends BaseTransformer {

    @Override
    public CompilationUnit transform(CompilationUnit cu) {

        ModifierVisitor<Void> transformer = new ModifierVisitor<>() {

            @Override
            public Visitable visit(MethodDeclaration md, Void arg) {
                if (md.getBody().isEmpty()) return md;

                BlockStmt body = md.getBody().get();
                BlockStmt newBody = removeConsecutiveDeadStores(body);
                md.setBody(newBody);

                return super.visit(md, arg);
            }

            private BlockStmt removeConsecutiveDeadStores(BlockStmt body) {
                List<Statement> statements = body.getStatements();
                List<Statement> newBody = new ArrayList<>();

                String lastVar = null;
                Statement lastStmt = null;

                for (Statement stmt : statements) {
                    if (stmt.isExpressionStmt()) {
                        Expression expr = stmt.asExpressionStmt().getExpression();

                        if (expr.isAssignExpr()) {
                            AssignExpr assign = expr.asAssignExpr();
                            String varName = extractTargetId(assign.getTarget());

                            if (varName != null) {
                                if (lastVar != null && lastVar.equals(varName) && newBody.contains(lastStmt)) {
                                    Set<String> usedNames = findUsedNames(assign);
                                    if (!usedNames.contains(lastVar)) {
                                        newBody.remove(lastStmt);
                                        logRule(String.format(
                                                "%s: Removed consecutive dead store to '%s' at line %d",
                                                this.getClass().getSimpleName(),
                                                varName,
                                                lastStmt.getBegin().map(p -> p.line).orElse(-1)
                                        ));
                                    }
                                }

                                lastVar = varName;
                                lastStmt = stmt;
                                newBody.add(stmt);
                                continue;
                            }
                        }
                    }

                    lastVar = null;
                    lastStmt = null;
                    newBody.add(stmt);
                }

                BlockStmt cleaned = new BlockStmt();
                newBody.forEach(cleaned::addStatement);
                return cleaned;
            }

            private Set<String> findUsedNames(Node node) {
                Set<String> used = new HashSet<>();

                node.walk(NameExpr.class, ne -> used.add(ne.getNameAsString()));
                node.walk(FieldAccessExpr.class, fae -> {
                    if (fae.getScope().isNameExpr()) {
                        used.add(fae.getScope().asNameExpr().getNameAsString() + "." + fae.getNameAsString());
                    }
                });

                return used;
            }

            private String extractTargetId(Expression target) {
                if (target.isNameExpr()) {
                    return target.asNameExpr().getNameAsString();
                }
                if (target.isFieldAccessExpr()) {
                    FieldAccessExpr fae = target.asFieldAccessExpr();
                    if (fae.getScope().isNameExpr()) {
                        return fae.getScope().asNameExpr().getNameAsString() + "." + fae.getNameAsString();
                    }
                }
                return null;
            }
        };

        transformer.visit(cu, null);
        return cu;
    }
}

File: greencoderefactor-python/trasnsformers/sonar/r1854/unusedassignmenttransformer.py
import ast
from trasnsformers.basetransformer import BaseTransformer

class UnusedAssignmentTransformer(BaseTransformer):
    """
    Sonar rule R1854.
    """

    def visit(self, node):
        for field, value in ast.iter_fields(node):
            if isinstance(value, list):
                new_list = []
                for item in value:
                    if isinstance(item, ast.stmt):
                        new_list.append(self.visit(item))
                    else:
                        new_list.append(item)
                setattr(node, field, self._remove_consecutive_dead_stores(new_list))
            elif isinstance(value, ast.AST):
                setattr(node, field, self.visit(value))
        return node

    def _remove_consecutive_dead_stores(self, stmts):
        new_body = []
        last_var = None
        last_stmt = None

        for stmt in stmts:
            if isinstance(stmt, ast.Assign) and len(stmt.targets) == 1:
                var_name = self._get_var_name(stmt.targets[0])
                if var_name:
                    if last_var == var_name and last_stmt in new_body:
                        used_in_current = self._find_used_names(stmt)
                        if last_var not in used_in_current:
                            new_body.remove(last_stmt)
                            self.log_rule(
                                f"{self.__class__.__name__}: Removed consecutive dead store to '{var_name}' at line {last_stmt.lineno}"
                            )

                    last_var = var_name
                    last_stmt = stmt
                else:
                    last_var = None
                    last_stmt = None
            else:
                last_var = None
                last_stmt = None

            new_body.append(stmt)

        return new_body

    def _get_var_name(self, target):
        if isinstance(target, ast.Name):
            return target.id
        elif isinstance(target, ast.Attribute) and isinstance(target.value, ast.Name):
            return f"{target.value.id}.{target.attr}"
        return None

    def _find_used_names(self, node):
        used = set()

        class Visitor(ast.NodeVisitor):
            def visit_Name(self, n):
                if isinstance(n.ctx, ast.Load):
                    used.add(n.id)

            def visit_Attribute(self, n):
                if isinstance(n.ctx, ast.Load) and isinstance(n.value, ast.Name):
                    used.add(f"{n.value.id}.{n.attr}")
                self.generic_visit(n)

        Visitor().visit(node)
        return used
